json-rpc/src/data.rs

```diff
@@ -0,0 +1,423 @@
+        AccountStateWithProofView, AccountView, BytesView, CurrencyInfoView, EventView,
+        EventWithProofView, MetadataView, StateProofView, TransactionView, TransactionsProofsView,
+        TransactionsWithProofsView,
+    },
+};
+use anyhow::{format_err, Result};
+use diem_crypto::{hash::CryptoHash, HashValue};
+use diem_types::{
+    account_address::AccountAddress,
+    account_config::{
+        diem_root_address, from_currency_code_string, resources::dual_attestation::Limit,
+        AccountResource,
+    },
+    account_state::AccountState,
+    chain_id::ChainId,
+    event::EventKey,
+    ledger_info::LedgerInfoWithSignatures,
+};
+use network::counters;
+use std::{
+    cmp::min,
+    convert::{TryFrom, TryInto},
+    sync::Arc,
+};
+use storage_interface::{DbReader, Order};
+
+pub fn get_account_state(
+    db: &Arc<dyn DbReader>,
+    address: AccountAddress,
+    version: u64,
+) -> Result<Option<AccountState>> {
+    if let Some(blob) = db
+        .get_account_state_with_proof_by_version(address, version)?
+        .0
+    {
+        Ok(Some(AccountState::try_from(&blob)?))
+    } else {
+        Ok(None)
+    }
+}
+
+/// Returns the blockchain metadata for a specified version. If no version is specified, default to
+/// returning the current blockchain metadata
+/// Can be used to verify that target Full Node is up-to-date
+pub fn get_metadata(
+    db: &Arc<dyn DbReader>,
+    ledger_version: u64,
+    chain_id: ChainId,
+    version: u64,
+) -> Result<MetadataView, JsonRpcError> {
+    let mut script_hash_allow_list: Option<Vec<HashValue>> = None;
+    let mut module_publishing_allowed: Option<bool> = None;
+    let mut diem_version: Option<u64> = None;
+    let mut dual_attestation_limit: Option<u64> = None;
+    if version == ledger_version {
+        if let Some(account) = get_account_state(db, diem_root_address(), version)? {
+            if let Some(vm_publishing_option) = account.get_vm_publishing_option()? {
+                script_hash_allow_list = Some(vm_publishing_option.script_allow_list);
+
+                module_publishing_allowed = Some(vm_publishing_option.is_open_module);
+            }
+            if let Some(v) = account.get_diem_version()? {
+                diem_version = Some(v.major)
+            }
+            if let Some(limit) = account.get_resource::<Limit>()? {
+                dual_attestation_limit = Some(limit.micro_xdx_limit)
+            }
+        }
+    }
+    Ok(MetadataView {
+        version,
+        accumulator_root_hash: db.get_accumulator_root_hash(version)?,
+        timestamp: db.get_block_timestamp(version)?,
+        chain_id: chain_id.id(),
+        script_hash_allow_list,
+        module_publishing_allowed,
+        diem_version,
+        dual_attestation_limit,
+    })
+}
+
+/// Returns account state (AccountView) by given address
+pub fn get_account(
+    db: &Arc<dyn DbReader>,
+    ledger_version: u64,
+    account_address: AccountAddress,
+    version: u64,
```

Comment:
Any particular reason you wanted these split out instead of reusing the `Param` structs? (Not saying one way is better than the other, just curious on your rationale)

Reply:
I wanted to separate them in case more parameters are added which shouldn't necessarily be exposed to the user (the Param structs are tied with the json RPC requests)